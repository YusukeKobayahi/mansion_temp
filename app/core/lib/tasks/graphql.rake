# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
