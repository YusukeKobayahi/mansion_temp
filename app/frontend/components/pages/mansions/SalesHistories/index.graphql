fragment MsLibraryMansionsSalesHistories on Mansion
@argumentDefinitions(historyCursor: { type: "String", defaultValue: "" }) {
  salesHistories(first: 4, after: $historyCursor) {
    edges {
      node {
        id
        price
        squarePrice
        saleOn
        exclusiveArea
        floorNumber
        layout
      }
    }
    pageInfo {
      startCursor
      endCursor
      hasPreviousPage
      hasNextPage
    }
  }
}
