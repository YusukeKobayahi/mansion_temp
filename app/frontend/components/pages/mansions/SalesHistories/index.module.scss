@import "~/styles/variables";
@import "~/styles/utils";

.container {
  width: 800px;
  margin: 50px auto 0;

  @include screenMd {
    width: auto;
    padding: 0 16px;
  }
}

%table {
  width: 100%;
  font-size: 16px;
  line-height: 1.4;
}

.tablePc {
  @extend %table;

  th {
    padding: 14px;
    text-align: center;
    background-color: #faf9f7;
    border-top: 1px solid #d0d0d0;
    border-bottom: 1px solid #d0d0d0;
    color: #646464;
  }

  td {
    padding: 14px;
    text-align: center;
    border-top: 1px solid #d0d0d0;
    border-bottom: 1px solid #d0d0d0;
    color: #646464;
  }
}

.tableSp {
  @extend %table;

  tr {
    border-top: 1px solid #d0d0d0;
    border-bottom: 1px solid #d0d0d0;
    font-size: 14px;
  }

  th {
    text-align: left;
    padding: 10px;
    background-color: #faf9f7;
    vertical-align: middle;
  }

  td {
    text-align: left;
    padding: 10px;
    line-height: 1.4;
  }

  &__price {
    font-size: 18px;
    font-weight: bold;
  }

  &__info {
    font-size: 12px;
  }
}

.showmoreContainer {
  margin-top: 20px;
  text-align: center;
}
