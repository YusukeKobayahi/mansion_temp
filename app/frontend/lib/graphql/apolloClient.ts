// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
