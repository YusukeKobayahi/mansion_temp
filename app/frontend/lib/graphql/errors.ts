export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
