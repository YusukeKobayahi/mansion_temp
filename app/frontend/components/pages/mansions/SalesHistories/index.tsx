import { format } from "date-fns";
import { MsLibraryMansionsSalesHistoriesFragment } from "~/graphql/generated";
import compact from "lodash/compact";
import zip from "lodash/zip";
import { useDeepCompareMemo } from "use-deep-compare";

import styles from "./index.module.scss";
import SidelineHeading from "~/components/commons/headings/SidelineHeading";
import ShowMoreBtn from "~/components/commons/buttons/ShowMoreBtn";
import { yenFormat } from "~/lib/utils";

export interface SalesHistoriesProps {
  data: MsLibraryMansionsSalesHistoriesFragment;
  onLoadMore: () => void;
  isLoadingMore: boolean;
  hasHistoryNextPage: boolean;
}

const SalesHistories: React.FC<SalesHistoriesProps> = ({
  data,
  onLoadMore,
  isLoadingMore,
  hasHistoryNextPage,
}: SalesHistoriesProps) => {
  const histories = compact(
    data?.salesHistories?.edges?.map((edge) =>
      edge !== null ? edge.node : null
    )
  );
  const ids = useDeepCompareMemo(() => histories.map((node) => node.id), [
    histories,
  ]);
  const dates = useDeepCompareMemo(() => {
    return histories.map((node) =>
      format(Date.parse(node.saleOn), "yyyy年MM月")
    );
  }, [histories]);
  const prices = useDeepCompareMemo(
    () => histories.map((node) => yenFormat(node.price)),
    [histories]
  );
  const layouts = useDeepCompareMemo(
    () => histories.map((node) => node.layout),
    [histories]
  );
  const areas = useDeepCompareMemo(
    () => histories.map((node) => `${node.exclusiveArea}㎡`),
    [histories]
  );
  const floorNumbers = useDeepCompareMemo(
    () =>
      histories.map((node) =>
        node.floorNumber ? `${node.floorNumber}階` : "-"
      ),
    [histories]
  );

  const resultsForPC = zip(ids, dates, prices, layouts, areas, floorNumbers);
  const resultsForSp = zip(
    ids,
    dates,
    prices,
    layouts,
    areas,
    floorNumbers
  ).map((result) => {
    return [
      result[0],
      result[1],
      result[2],
      [result[3], result[4], result[5]].join(" / "),
    ];
  });

  return (
    <section className={styles.container}>
      <h3>
        <SidelineHeading>販売履歴</SidelineHeading>
      </h3>
      {resultsForPC.length && (
        <div className="pcOnly">
          <table className={styles.tablePc}>
            <thead>
              <tr>
                <th>年月</th>
                <th>価格</th>
                <th>間取り</th>
                <th>専有面積</th>
                <th>所在階</th>
              </tr>
            </thead>
            <tbody>
              {resultsForPC.map((arr) => (
                <tr key={arr[0]}>
                  <td>{arr[1]}</td>
                  <td>{arr[2]}</td>
                  <td>{arr[3]}</td>
                  <td>{arr[4]}</td>
                  <td>{arr[5]}</td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      )}
      {resultsForSp.length && (
        <div className="spOnly">
          <table className={styles.tableSp}>
            <tbody>
              {resultsForSp.map((arr) => (
                <tr key={arr[0]}>
                  <th>{arr[1]}</th>
                  <td>
                    <p className={styles.tableSp__price}>{arr[2]}</p>
                    <span className={styles.tableSp__info}>{arr[3]}</span>
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      )}
      {hasHistoryNextPage ? (
        <div className={styles.showmoreContainer}>
          <ShowMoreBtn onClick={onLoadMore} isLoading={isLoadingMore} />
        </div>
      ) : null}
    </section>
  );
};

export default SalesHistories;
