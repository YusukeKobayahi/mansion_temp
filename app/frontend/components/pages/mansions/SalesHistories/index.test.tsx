import { screen, render } from "@testing-library/react";
import mockData from "../fixtures/mockData.json";
import { MansionQuery } from "~/graphql/generated";
import SalesHistories from ".";

describe("components/pages/mansions/SalesHistories", () => {
  it("正常に表示される", () => {
    const mock = mockData as MansionQuery;
    const { asFragment } = render(
      <SalesHistories data={mock.mansion} hasHistoryNextPage={true} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("floorNumberがnullの場合に「null」という文字が表示されない", () => {
    const mock = mockData as MansionQuery;
    const firstHistory = mock.mansion.salesHistories.edges?.[0]?.node;
    if (!firstHistory) {
      throw "Invalid mockData";
    }
    firstHistory.floorNumber = null;

    render(<SalesHistories data={mock.mansion} hasHistoryNextPage={true} />);
    expect(screen.queryAllByText(/null/)).toHaveLength(0);
  });
});
