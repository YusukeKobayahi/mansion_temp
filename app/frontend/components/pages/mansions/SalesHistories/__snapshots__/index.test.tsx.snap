// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/mansions/SalesHistories 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <h3>
      <div
        class="heading"
      >
        <span
          class="inner"
        >
          販売履歴
        </span>
      </div>
    </h3>
    <div
      class="pcOnly"
    >
      <table
        class="tablePc"
      >
        <thead>
          <tr>
            <th>
              年月
            </th>
            <th>
              価格
            </th>
            <th>
              間取り
            </th>
            <th>
              専有面積
            </th>
            <th>
              所在階
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              2019年05月
            </td>
            <td>
              2,496万円
            </td>
            <td>
              1K
            </td>
            <td>
              22.72㎡
            </td>
            <td>
              9階
            </td>
          </tr>
          <tr>
            <td>
              2019年05月
            </td>
            <td>
              2,590万円
            </td>
            <td>
              1K
            </td>
            <td>
              22.72㎡
            </td>
            <td>
              9階
            </td>
          </tr>
          <tr>
            <td>
              2019年05月
            </td>
            <td>
              4,550万円
            </td>
            <td>
              1DK
            </td>
            <td>
              45.16㎡
            </td>
            <td>
              10階
            </td>
          </tr>
          <tr>
            <td>
              2019年05月
            </td>
            <td>
              4,650万円
            </td>
            <td>
              1DK
            </td>
            <td>
              45.16㎡
            </td>
            <td>
              10階
            </td>
          </tr>
          <tr>
            <td>
              2019年03月
            </td>
            <td>
              3,200万円
            </td>
            <td>
              1K
            </td>
            <td>
              30.24㎡
            </td>
            <td>
              11階
            </td>
          </tr>
          <tr>
            <td>
              2019年03月
            </td>
            <td>
              2,590万円
            </td>
            <td>
              1K
            </td>
            <td>
              22.72㎡
            </td>
            <td>
              9階
            </td>
          </tr>
          <tr>
            <td>
              2019年02月
            </td>
            <td>
              4,580万円
            </td>
            <td>
              1DK
            </td>
            <td>
              45.16㎡
            </td>
            <td>
              4階
            </td>
          </tr>
          <tr>
            <td>
              2019年02月
            </td>
            <td>
              4,780万円
            </td>
            <td>
              1DK
            </td>
            <td>
              45.16㎡
            </td>
            <td>
              4階
            </td>
          </tr>
          <tr>
            <td>
              2018年09月
            </td>
            <td>
              2,000万円
            </td>
            <td>
              1K
            </td>
            <td>
              20.72㎡
            </td>
            <td>
              3階
            </td>
          </tr>
          <tr>
            <td>
              2018年09月
            </td>
            <td>
              3,480万円
            </td>
            <td>
              1DK
            </td>
            <td>
              45.16㎡
            </td>
            <td>
              2階
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="spOnly"
    >
      <table
        class="tableSp"
      >
        <tbody>
          <tr>
            <th>
              2019年05月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                2,496万円
              </p>
              <span
                class="tableSp__info"
              >
                1K / 22.72㎡ / 9階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2019年05月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                2,590万円
              </p>
              <span
                class="tableSp__info"
              >
                1K / 22.72㎡ / 9階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2019年05月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                4,550万円
              </p>
              <span
                class="tableSp__info"
              >
                1DK / 45.16㎡ / 10階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2019年05月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                4,650万円
              </p>
              <span
                class="tableSp__info"
              >
                1DK / 45.16㎡ / 10階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2019年03月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                3,200万円
              </p>
              <span
                class="tableSp__info"
              >
                1K / 30.24㎡ / 11階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2019年03月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                2,590万円
              </p>
              <span
                class="tableSp__info"
              >
                1K / 22.72㎡ / 9階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2019年02月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                4,580万円
              </p>
              <span
                class="tableSp__info"
              >
                1DK / 45.16㎡ / 4階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2019年02月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                4,780万円
              </p>
              <span
                class="tableSp__info"
              >
                1DK / 45.16㎡ / 4階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2018年09月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                2,000万円
              </p>
              <span
                class="tableSp__info"
              >
                1K / 20.72㎡ / 3階
              </span>
            </td>
          </tr>
          <tr>
            <th>
              2018年09月
            </th>
            <td>
              <p
                class="tableSp__price"
              >
                3,480万円
              </p>
              <span
                class="tableSp__info"
              >
                1DK / 45.16㎡ / 2階
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="showmoreContainer"
    >
      <button
        class="btn"
      >
        さらに表示する
      </button>
    </div>
  </section>
</DocumentFragment>
`;
